### PR TITLE
Instance Segmentation Head

### PR DESCRIPTION
**Pull request by JSabadin**
# New Instance Segmentation Loss and Related Enhancements

## Description

In this PR, we propose:

- **New Detection Head**: Enhanced architecture for improved performance.
- **New Segmentation Head**: Built on top of the detection head to support instance segmentation.
- **New Loss Functions**:
  - Detection loss: Optimized for precision.
  - Segmentation loss: Designed for instance segmentation.
- **New Visualizer**: Provides visualization of instance segmentation training results.
- **Extended mAP Metric**: Now supports instance segmentation evaluation.
- **Luxonis-ML Loader Support**: Added compatibility with the new loader.

### Overfitting Examples

example2
example1
example3

### CODE REVIEW

File: luxonis_train/config/predefined_models/instance_segmentation_model.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    ModelNodeConfig,
+    Params,
+)
+
+from .base_predefined_model import BasePredefinedModel
+
+VariantLiteral: TypeAlias = Literal["light", "medium", "heavy"]
+
+
+class DetectionVariant(BaseModel):
```
Comment (klemen1999) [inline luxonis_train/config/predefined_models/instance_segmentation_model.py:18]: Rename to e.g. `InstanceSegmentationVariant`

File: luxonis_train/config/predefined_models/instance_segmentation_model.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    backbone: str
+    backbone_params: Params
+    neck_params: Params
+
+
+def get_variant(variant: VariantLiteral) -> DetectionVariant:
+    """Returns the specific variant configuration for the
+    DetectionModel."""
```
Comment (klemen1999) [inline luxonis_train/config/predefined_models/instance_segmentation_model.py:26]: Update docstrings accordingly

File: luxonis_train/config/predefined_models/instance_segmentation_model.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        nodes.append(
+            ModelNodeConfig(
+                name="PrecisionSegmentBBoxHead",
+                alias=f"{self.task_name}/PrecisionSegmentBBoxHead",
+                freezing=self.head_params.pop("freezing", {}),
+                inputs=[f"{self.task_name}/RepPANNeck"]
+                if self.use_neck
+                else [f"{self.backbone}-{self.task_name}"],
+                params=self.head_params,
+                # task=self.task_name,
```
Comment (klemen1999) [inline luxonis_train/config/predefined_models/instance_segmentation_model.py:117]: Can be removed

**Review by klemen1999 — APPROVED** [review]
Generally LGTM, left some small comments
